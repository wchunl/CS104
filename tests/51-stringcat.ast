ROOT "" (0.0.0)
|   TOK_FUNCTION "" (6.22.1)
|   |   TOK_TYPE_ID "" (6.22.1)
|   |   |   TOK_VOID "void" (6.22.1)
|   |   |   TOK_IDENT "fail" (6.22.6)
|   |   TOK_PARAM "(" (6.22.11)
|   |   |   TOK_TYPE_ID "" (6.22.12)
|   |   |   |   TOK_STRING "string" (6.22.12)
|   |   |   |   TOK_IDENT "expr" (6.22.19)
|   |   |   TOK_TYPE_ID "" (6.22.25)
|   |   |   |   TOK_STRING "string" (6.22.25)
|   |   |   |   TOK_IDENT "file" (6.22.32)
|   |   |   TOK_TYPE_ID "" (6.22.38)
|   |   |   |   TOK_INT "int" (6.22.38)
|   |   |   |   TOK_IDENT "line" (6.22.42)
|   TOK_FUNCTION "" (6.24.1)
|   |   TOK_TYPE_ID "" (6.24.1)
|   |   |   TOK_VOID "void" (6.24.1)
|   |   |   TOK_IDENT "putchr" (6.24.6)
|   |   TOK_PARAM "(" (6.24.13)
|   |   |   TOK_TYPE_ID "" (6.24.14)
|   |   |   |   TOK_INT "int" (6.24.14)
|   |   |   |   TOK_IDENT "chr" (6.24.18)
|   TOK_FUNCTION "" (6.25.1)
|   |   TOK_TYPE_ID "" (6.25.1)
|   |   |   TOK_VOID "void" (6.25.1)
|   |   |   TOK_IDENT "putint" (6.25.6)
|   |   TOK_PARAM "(" (6.25.13)
|   |   |   TOK_TYPE_ID "" (6.25.14)
|   |   |   |   TOK_INT "int" (6.25.14)
|   |   |   |   TOK_IDENT "num" (6.25.18)
|   TOK_FUNCTION "" (6.26.1)
|   |   TOK_TYPE_ID "" (6.26.1)
|   |   |   TOK_VOID "void" (6.26.1)
|   |   |   TOK_IDENT "putstr" (6.26.6)
|   |   TOK_PARAM "(" (6.26.13)
|   |   |   TOK_TYPE_ID "" (6.26.14)
|   |   |   |   TOK_STRING "string" (6.26.14)
|   |   |   |   TOK_IDENT "str" (6.26.21)
|   TOK_FUNCTION "" (6.28.1)
|   |   TOK_TYPE_ID "" (6.28.1)
|   |   |   TOK_INT "int" (6.28.1)
|   |   |   TOK_IDENT "getchr" (6.28.5)
|   |   TOK_PARAM "(" (6.28.11)
|   TOK_FUNCTION "" (6.29.1)
|   |   TOK_TYPE_ID "" (6.29.1)
|   |   |   TOK_STRING "string" (6.29.1)
|   |   |   TOK_IDENT "getstr" (6.29.8)
|   |   TOK_PARAM "(" (6.29.14)
|   TOK_FUNCTION "" (6.30.1)
|   |   TOK_TYPE_ID "" (6.30.1)
|   |   |   TOK_STRING "string" (6.30.1)
|   |   |   TOK_IDENT "getln" (6.30.8)
|   |   TOK_PARAM "(" (6.30.13)
|   TOK_FUNCTION "" (7.8.1)
|   |   TOK_TYPE_ID "" (7.8.1)
|   |   |   TOK_INT "int" (7.8.1)
|   |   |   TOK_IDENT "strlen" (7.8.5)
|   |   TOK_PARAM "(" (7.8.12)
|   |   |   TOK_TYPE_ID "" (7.8.13)
|   |   |   |   TOK_STRING "string" (7.8.13)
|   |   |   |   TOK_IDENT "str" (7.8.20)
|   |   TOK_BLOCK "{" (7.8.25)
|   |   |   TOK_FUNCTION "" (7.9.12)
|   |   |   |   TOK_VARDECL "=" (7.9.12)
|   |   |   |   |   TOK_INT "int" (7.9.4)
|   |   |   |   |   |   TOK_TYPE_ID "len" (7.9.8)
|   |   |   |   |   TOK_INTCON "0" (7.9.14)
|   |   |   |   TOK_WHILE "while" (7.10.4)
|   |   |   |   |   TOK_NE "!=" (7.10.20)
|   |   |   |   |   |   TOK_INDEX "[" (7.10.14)
|   |   |   |   |   |   |   TOK_IDENT "str" (7.10.11)
|   |   |   |   |   |   |   TOK_IDENT "len" (7.10.15)
|   |   |   |   |   |   TOK_CHARCON "'\0'" (7.10.23)
|   |   |   |   |   '=' "=" (7.10.33)
|   |   |   |   |   |   TOK_IDENT "len" (7.10.29)
|   |   |   |   |   |   '+' "+" (7.10.39)
|   |   |   |   |   |   |   TOK_IDENT "len" (7.10.35)
|   |   |   |   |   |   |   TOK_INTCON "1" (7.10.41)
|   |   |   |   TOK_RETURN "return" (7.11.4)
|   |   |   |   |   TOK_IDENT "len" (7.11.11)
|   TOK_FUNCTION "" (7.14.1)
|   |   TOK_TYPE_ID "" (7.14.1)
|   |   |   TOK_VOID "void" (7.14.1)
|   |   |   TOK_IDENT "strcat" (7.14.6)
|   |   TOK_PARAM "(" (7.14.13)
|   |   |   TOK_TYPE_ID "" (7.14.14)
|   |   |   |   TOK_STRING "string" (7.14.14)
|   |   |   |   TOK_IDENT "dest" (7.14.21)
|   |   |   TOK_TYPE_ID "" (7.14.27)
|   |   |   |   TOK_STRING "string" (7.14.27)
|   |   |   |   TOK_IDENT "src" (7.14.34)
|   |   TOK_BLOCK "{" (7.14.39)
|   |   |   TOK_FUNCTION "" (7.15.12)
|   |   |   |   TOK_VARDECL "=" (7.15.12)
|   |   |   |   |   TOK_INT "int" (7.15.4)
|   |   |   |   |   |   TOK_TYPE_ID "pos" (7.15.8)
|   |   |   |   |   TOK_CALL "(" (7.15.21)
|   |   |   |   |   |   TOK_IDENT "strlen" (7.15.14)
|   |   |   |   |   |   TOK_IDENT "dest" (7.15.22)
|   |   |   |   TOK_VARDECL "=" (7.16.14)
|   |   |   |   |   TOK_INT "int" (7.16.4)
|   |   |   |   |   |   TOK_TYPE_ID "srcix" (7.16.8)
|   |   |   |   |   TOK_INTCON "0" (7.16.16)
|   |   |   |   TOK_WHILE "while" (7.17.4)
|   |   |   |   |   TOK_NE "!=" (7.17.22)
|   |   |   |   |   |   TOK_INDEX "[" (7.17.14)
|   |   |   |   |   |   |   TOK_IDENT "src" (7.17.11)
|   |   |   |   |   |   |   TOK_IDENT "srcix" (7.17.15)
|   |   |   |   |   |   TOK_CHARCON "'\0'" (7.17.25)
|   |   |   |   |   TOK_BLOCK "{" (7.17.31)
|   |   |   |   |   |   TOK_FUNCTION "" (7.18.17)
|   |   |   |   |   |   |   '=' "=" (7.18.17)
|   |   |   |   |   |   |   |   TOK_INDEX "[" (7.18.11)
|   |   |   |   |   |   |   |   |   TOK_IDENT "dest" (7.18.7)
|   |   |   |   |   |   |   |   |   TOK_IDENT "pos" (7.18.12)
|   |   |   |   |   |   |   |   TOK_INDEX "[" (7.18.22)
|   |   |   |   |   |   |   |   |   TOK_IDENT "src" (7.18.19)
|   |   |   |   |   |   |   |   |   TOK_IDENT "srcix" (7.18.23)
|   |   |   |   |   |   |   '=' "=" (7.19.11)
|   |   |   |   |   |   |   |   TOK_IDENT "pos" (7.19.7)
|   |   |   |   |   |   |   |   '+' "+" (7.19.17)
|   |   |   |   |   |   |   |   |   TOK_IDENT "pos" (7.19.13)
|   |   |   |   |   |   |   |   |   TOK_INTCON "1" (7.19.19)
|   |   |   |   |   |   |   '=' "=" (7.20.13)
|   |   |   |   |   |   |   |   TOK_IDENT "srcix" (7.20.7)
|   |   |   |   |   |   |   |   '+' "+" (7.20.21)
|   |   |   |   |   |   |   |   |   TOK_IDENT "srcix" (7.20.15)
|   |   |   |   |   |   |   |   |   TOK_INTCON "1" (7.20.23)
|   |   |   |   '=' "=" (7.22.14)
|   |   |   |   |   TOK_INDEX "[" (7.22.8)
|   |   |   |   |   |   TOK_IDENT "dest" (7.22.4)
|   |   |   |   |   |   TOK_IDENT "pos" (7.22.9)
|   |   |   |   |   TOK_CHARCON "'\0'" (7.22.16)
|   TOK_FUNCTION "" (7.25.1)
|   |   TOK_TYPE_ID "" (7.25.1)
|   |   |   TOK_INT "int" (7.25.1)
|   |   |   TOK_IDENT "main" (7.25.5)
|   |   TOK_PARAM "(" (7.25.10)
|   |   |   TOK_TYPE_ID "" (7.25.11)
|   |   |   |   TOK_INT "int" (7.25.11)
|   |   |   |   TOK_IDENT "argc" (7.25.15)
|   |   |   TOK_TYPE_ID "" (7.25.21)
|   |   |   |   TOK_ARRAY "array" (7.25.21)
|   |   |   |   |   TOK_TYPE_ID "string" (7.25.27)
|   |   |   |   TOK_IDENT "argv" (7.25.35)
|   |   TOK_BLOCK "{" (7.25.41)
|   |   |   TOK_FUNCTION "" (7.26.15)
|   |   |   |   TOK_VARDECL "=" (7.26.15)
|   |   |   |   |   TOK_INT "int" (7.26.4)
|   |   |   |   |   |   TOK_TYPE_ID "length" (7.26.8)
|   |   |   |   |   TOK_INTCON "0" (7.26.17)
|   |   |   |   TOK_VARDECL "=" (7.27.13)
|   |   |   |   |   TOK_INT "int" (7.27.4)
|   |   |   |   |   |   TOK_TYPE_ID "argi" (7.27.8)
|   |   |   |   |   TOK_INTCON "1" (7.27.15)
|   |   |   |   TOK_WHILE "while" (7.28.4)
|   |   |   |   |   '<' "<" (7.28.16)
|   |   |   |   |   |   TOK_IDENT "argi" (7.28.11)
|   |   |   |   |   |   TOK_IDENT "argc" (7.28.18)
|   |   |   |   |   TOK_BLOCK "{" (7.28.24)
|   |   |   |   |   |   TOK_FUNCTION "" (7.29.14)
|   |   |   |   |   |   |   '=' "=" (7.29.14)
|   |   |   |   |   |   |   |   TOK_IDENT "length" (7.29.7)
|   |   |   |   |   |   |   |   '+' "+" (7.29.45)
|   |   |   |   |   |   |   |   |   '+' "+" (7.29.23)
|   |   |   |   |   |   |   |   |   |   TOK_IDENT "length" (7.29.16)
|   |   |   |   |   |   |   |   |   |   TOK_CALL "(" (7.29.32)
|   |   |   |   |   |   |   |   |   |   |   TOK_IDENT "strlen" (7.29.25)
|   |   |   |   |   |   |   |   |   |   |   TOK_INDEX "[" (7.29.37)
|   |   |   |   |   |   |   |   |   |   |   |   TOK_IDENT "argv" (7.29.33)
|   |   |   |   |   |   |   |   |   |   |   |   TOK_IDENT "argi" (7.29.38)
|   |   |   |   |   |   |   |   |   TOK_INTCON "2" (7.29.47)
|   |   |   |   |   |   |   '=' "=" (7.30.12)
|   |   |   |   |   |   |   |   TOK_IDENT "argi" (7.30.7)
|   |   |   |   |   |   |   |   '+' "+" (7.30.19)
|   |   |   |   |   |   |   |   |   TOK_IDENT "argi" (7.30.14)
|   |   |   |   |   |   |   |   |   TOK_INTCON "1" (7.30.21)
|   |   |   |   TOK_CALL "(" (7.32.11)
|   |   |   |   |   TOK_IDENT "putstr" (7.32.4)
|   |   |   |   |   TOK_STRINGCON ""concat length = "" (7.32.12)
|   |   |   |   TOK_CALL "(" (7.33.11)
|   |   |   |   |   TOK_IDENT "putint" (7.33.4)
|   |   |   |   |   TOK_IDENT "length" (7.33.12)
|   |   |   |   TOK_CALL "(" (7.34.11)
|   |   |   |   |   TOK_IDENT "putchr" (7.34.4)
|   |   |   |   |   TOK_CHARCON "'\n'" (7.34.12)
|   |   |   |   TOK_VARDECL "=" (7.35.18)
|   |   |   |   |   TOK_STRING "string" (7.35.4)
|   |   |   |   |   |   TOK_TYPE_ID "concat" (7.35.11)
|   |   |   |   |   TOK_ALLOC "alloc" (7.35.20)
|   |   |   |   |   |   TOK_STRING "string" (7.35.26)
|   |   |   |   |   |   TOK_IDENT "length" (7.35.35)
|   |   |   |   '=' "=" (7.36.9)
|   |   |   |   |   TOK_IDENT "argi" (7.36.4)
|   |   |   |   |   TOK_INTCON "1" (7.36.11)
|   |   |   |   TOK_CALL "(" (7.37.11)
|   |   |   |   |   TOK_IDENT "putchr" (7.37.4)
|   |   |   |   |   TOK_CHARCON "'['" (7.37.12)
|   |   |   |   TOK_WHILE "while" (7.38.4)
|   |   |   |   |   '<' "<" (7.38.16)
|   |   |   |   |   |   TOK_IDENT "argi" (7.38.11)
|   |   |   |   |   |   TOK_IDENT "argc" (7.38.18)
|   |   |   |   |   TOK_BLOCK "{" (7.38.24)
|   |   |   |   |   |   TOK_FUNCTION "" (7.39.14)
|   |   |   |   |   |   |   TOK_CALL "(" (7.39.14)
|   |   |   |   |   |   |   |   TOK_IDENT "strcat" (7.39.7)
|   |   |   |   |   |   |   |   TOK_IDENT "concat" (7.39.15)
|   |   |   |   |   |   |   |   |   TOK_STRINGCON ""("" (7.39.23)
|   |   |   |   |   |   |   TOK_CALL "(" (7.40.14)
|   |   |   |   |   |   |   |   TOK_IDENT "strcat" (7.40.7)
|   |   |   |   |   |   |   |   TOK_IDENT "concat" (7.40.15)
|   |   |   |   |   |   |   |   |   TOK_INDEX "[" (7.40.27)
|   |   |   |   |   |   |   |   |   |   TOK_IDENT "argv" (7.40.23)
|   |   |   |   |   |   |   |   |   |   TOK_IDENT "argi" (7.40.28)
|   |   |   |   |   |   |   TOK_CALL "(" (7.41.14)
|   |   |   |   |   |   |   |   TOK_IDENT "strcat" (7.41.7)
|   |   |   |   |   |   |   |   TOK_IDENT "concat" (7.41.15)
|   |   |   |   |   |   |   |   |   TOK_STRINGCON "")"" (7.41.23)
|   |   |   |   |   |   |   '=' "=" (7.42.12)
|   |   |   |   |   |   |   |   TOK_IDENT "argi" (7.42.7)
|   |   |   |   |   |   |   |   '+' "+" (7.42.19)
|   |   |   |   |   |   |   |   |   TOK_IDENT "argi" (7.42.14)
|   |   |   |   |   |   |   |   |   TOK_INTCON "1" (7.42.21)
|   |   |   |   TOK_CALL "(" (7.44.11)
|   |   |   |   |   TOK_IDENT "putstr" (7.44.4)
|   |   |   |   |   TOK_IDENT "concat" (7.44.12)
|   |   |   |   TOK_CALL "(" (7.45.11)
|   |   |   |   |   TOK_IDENT "putstr" (7.45.4)
|   |   |   |   |   TOK_STRINGCON ""]\n"" (7.45.12)
|   |   |   |   TOK_RETURN "return" (7.46.4)
|   |   |   |   |   TOK_INTCON "0" (7.46.11)
