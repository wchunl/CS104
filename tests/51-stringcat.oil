fail:     .function
          .param ptr expr
          .param ptr file
          .param int line
          .end
putchr:   .function
          .param int chr
          .end
putint:   .function
          .param int num
          .end
putstr:   .function
          .param ptr str
          .end
getchr:   .function int
          .end
getstr:   .function
          .end
getln:    .function
          .end
strlen:   .function int
          .param ptr str
          .local int len
          len = 0
.wh0:     $t0:i = str[len * :p]
          $t1:i = $t0:i != '\0'
          goto .od0 if not $t1:i
.do0:     $t2:i = len + 1
          len = $t2:i
.od0:     return len
          return 
          .end
strcat:   .function
          .param ptr dest
          .param ptr src
          .local int pos
          .local int srcix
          strlen(dest)
          pos = $t3:i
          srcix = 0
.wh1:     $t4:i = src[srcix * :p]
          $t5:i = $t4:i != '\0'
          goto .od1 if not $t5:i
.do1:     $t6:p = src[srcix * :p]
          dest[pos * :p] = $t6:i
          $t7:i = pos + 1
          pos = $t7:i
          $t8:i = srcix + 1
          srcix = $t8:i
.od1:     dest[pos * :p] = '\0'
          .end
main:     .function int
          .param int argc
          .param ptr argv
          .local int length
          .local int argi
          .local ptr concat
          length = 0
          argi = 1
.wh2:     $t9:i = argi < argc
          goto .od2 if not $t9:i
.do2:     $t10:i = + + 2
          length = $t10:i
          $t11:i = argi + 1
          argi = $t11:i
.od2:     putstr("concat length = ")
          putint(length)
          putchr('\n')
          malloc(4)
          concat = $t12:i
          argi = 1
          putchr('[')
.wh3:     $t13:i = argi < argc
          goto .od3 if not $t13:i
.do3:     $t14:i = concat
          strcat($t14:i)
          $t15:i = concat
          strcat($t15:i)
          $t16:i = concat
          strcat($t16:i)
          $t17:i = argi + 1
          argi = $t17:i
.od3:     putstr(concat)
          putstr("]\n")
          return 0
          return 
          .end
