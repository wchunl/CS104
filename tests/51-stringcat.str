string_set[   2]:   12597418120160561823 0xe6bb48->"int"
string_set[   3]:   15410491504089378546 0xe70458->""concat length = ""
                   11752130776608228127 0xe6ddb8->"'\0'"
string_set[   5]:    1370730858159036685 0xe6de58->"+"
string_set[   7]:    3505610083221236292 0xe6fab8->"length"
string_set[   8]:    4551451650890805270 0xe6d798->"2"
                   13497578857207247412 0xe6ba08->"string"
string_set[   9]:   10319632053171904911 0xe6da78->"while"
string_set[  10]:    9201800968376702500 0xe6cc38->"getchr"
string_set[  12]:    1363220277528623646 0xe6d2b8->"strlen"
                   14643508047410943861 0xe6fdc8->"argv"
string_set[  14]:   15816506932648033945 0xe6b968->"("
                    1396094175042499165 0xe6fc68->"<"
string_set[  15]:   12868445110721718657 0xe6bd18->")"
string_set[  17]:    6142509188972423790 0xe675e8->""
string_set[  20]:    1583576014053361350 0xe6c888->"str"
                    9423661692722584567 0xe6e218->"dest"
string_set[  22]:   10159970873491820195 0xe6e058->"1"
string_set[  23]:   17897127611569151174 0xe71af8->"")""
string_set[  25]:    2178305242944497711 0xe6eaf8->"srcix"
string_set[  26]:   11597697714117577063 0xe6d8a8->"="
                   12291392672316406109 0xe6bea8->"line"
                    9611428142580219387 0xe67608->"fail"
string_set[  28]:   18413295556186398326 0xe6c0f8->"putint"
string_set[  29]:   11911796782241914524 0xe67218->"void"
string_set[  33]:    6814028903552918164 0xe6fbc8->"array"
string_set[  35]:   11868120863286193964 0xe6f818->"main"
string_set[  36]:   15761577654341137060 0xe72148->""]\n""
string_set[  41]:   16299098330955776099 0xe71648->""(""
                    3012229484421721799 0xe714a8->"'['"
                   12051031859636400188 0xe6dca8->"]"
string_set[  42]:   13686721134187162920 0xe6e0f8->"}"
string_set[  49]:   12679825168387489523 0xe6c708->"num"
                   13310716766371036271 0xe6c038->"len"
string_set[  50]:      21606623943142546 0xe6d638->"{"
string_set[  52]:    5344662657899890615 0xe6bec8->";"
string_set[  54]:    7296690267450279874 0xe6c368->"chr"
string_set[  55]:    2297668033614959926 0xe6d948->"0"
string_set[  56]:    3701972582333163920 0xe6f7d8->"argc"
string_set[  59]:    6192799629375248351 0xe70d88->"'\n'"
                   16541611849048157824 0xe6e4e8->"pos"
string_set[  63]:   14526079891305879776 0xe6fd58->">"
string_set[  64]:   16579916199511856337 0xe6d098->"getln"
string_set[  67]:    4706206694200543916 0xe6bc08->","
string_set[  69]:    5266873771696973564 0xe702f8->"argi"
string_set[  73]:   11100957978370531577 0xe6e648->"src"
                    9386170080689500517 0xe6bad8->"expr"
string_set[  75]:   15141261116119706963 0xe710f8->"concat"
string_set[  76]:   11166701219533395393 0xe6ce58->"getstr"
string_set[  80]:    8290819294854170237 0xe71278->"alloc"
                    7211130878751977441 0xe6c228->"putchr"
                    2617056156435215158 0xe6dcc8->"!="
string_set[  83]:   13581035385482130682 0xe6dbb8->"["
string_set[  85]:    4017852018407011111 0xe6bcf8->"file"
string_set[  87]:   14635384904416252381 0xe6d9e8->"return"
string_set[  89]:    5905351119970222309 0xe6e258->"strcat"
                   13348923631108115385 0xe6c4e8->"putstr"
load_factor = 0.588
bucket_count = 97
max_bucket_size = 3
