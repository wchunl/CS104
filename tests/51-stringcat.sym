fail (6.22.1) {0} void function
   expr (6.22.12) {1} string variable param lval 0
   file (6.22.25) {1} string variable param lval 1
   line (6.22.38) {1} int variable param lval 2

putchr (6.24.1) {1} void function
   chr (6.24.14) {1} int variable param lval 0

putint (6.25.1) {2} void function
   num (6.25.14) {1} int variable param lval 0

putstr (6.26.1) {3} void function
   str (6.26.14) {1} string variable param lval 0

getchr (6.28.1) {4} int function

getstr (6.29.1) {5} string function

getln (6.30.1) {6} string function

strlen (7.8.1) {7} int function
   str (7.8.13) {1} string variable param lval 0
   len (7.9.12) {1} int variable local lval 0

strcat (7.14.1) {8} void function
   dest (7.14.14) {1} string variable param lval 0
   src (7.14.27) {1} string variable param lval 1
   pos (7.15.12) {1} int variable local lval 0
   srcix (7.16.14) {1} int variable local lval 1

main (7.25.1) {9} int function
   argc (7.25.11) {1} int variable param lval 0
   argv (7.25.21) {1} array <string> variable param lval 1
   length (7.26.15) {1} int variable local lval 0
   argi (7.27.13) {1} int variable local lval 1
   concat (7.35.18) {1} string variable local lval 2
