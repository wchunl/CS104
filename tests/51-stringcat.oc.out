; 0.  1: # 1 "51-stringcat.oc"; 1.  0: 
; 1.  1: # 1 "<built-in>"
; 2.  1: # 1 "<command-line>"
; 3.  1: # 1 "51-stringcat.oc"
; 4.  1: 
; 4.  2: 
; 4.  3: 
; 4.  4: 
; 4.  5: 
; 4.  6: # 1 "oclib.h" 1
; 5.  1: # 22 "oclib.h"
; 6. 22: void fail (string expr, string file, int line);
; 6. 23: 
; 6. 24: void putchr (int chr);
; 6. 25: void putint (int num);
; 6. 26: void putstr (string str);
; 6. 27: 
; 6. 28: int getchr();
; 6. 29: string getstr();
; 6. 30: string getln();
; 6. 31: # 7 "51-stringcat.oc" 2
; 7.  7: 
; 7.  8: int strlen (string str) {
; 7.  9:    int len = 0;
; 7. 10:    while (str[len] != '\0') len = len + 1;
; 7. 11:    return len;
; 7. 12: }
; 7. 13: 
; 7. 14: void strcat (string dest, string src) {
; 7. 15:    int pos = strlen (dest);
; 7. 16:    int srcix = 0;
; 7. 17:    while (src[srcix] != '\0') {
; 7. 18:       dest[pos] = src[srcix];
; 7. 19:       pos = pos + 1;
; 7. 20:       srcix = srcix + 1;
; 7. 21:    }
; 7. 22:    dest[pos] = '\0';
; 7. 23: }
; 7. 24: 
; 7. 25: int main (int argc, array<string> argv) {
; 7. 26:    int length = 0;
; 7. 27:    int argi = 1;
; 7. 28:    while (argi < argc) {
; 7. 29:       length = length + strlen (argv[argi]) + 2;
; 7. 30:       argi = argi + 1;
; 7. 31:    }
; 7. 32:    putstr ("concat length = ");
; 7. 33:    putint (length);
; 7. 34:    putchr ('\n');
; 7. 35:    string concat = alloc<string> (length);
; 7. 36:    argi = 1;
; 7. 37:    putchr ('[');
; 7. 38:    while (argi < argc) {
; 7. 39:       strcat (concat, "(");
; 7. 40:       strcat (concat, argv[argi]);
; 7. 41:       strcat (concat, ")");
; 7. 42:       argi = argi + 1;
; 7. 43:    }
; 7. 44:    putstr (concat);
; 7. 45:    putstr ("]\n");
; 7. 46:    return 0;
; 7. 47: }
hello